File: src/kernel/archs/x86_64/paging.c
#include <hal>
#include <handover>
#include <logger>
#include <pmm>
#include <result>
#include <stddef.h>
#include <stdint.h>
#include <string.h>
#include <utils.h>

#include "asm.h"
#include "cpuid.h"
#include "paging.h"

static size_t page_size = mib$(2);
static uintptr_t *pml4 = NULL;

extern char text_start_addr[];
extern char text_end_addr[];
extern char rodata_start_addr[];
extern char rodata_end_addr[];
extern char data_start_addr[];
extern char data_end_addr[];

static int64_t transform_flags(HalMemFlags flags)
{
    int64_t ret_flags = PAGE_NO_EXECUTE | PAGE_PRESENT;

    if (flags & HAL_MEM_READ)
    {
    }

    if (flags & HAL_MEM_NONE)
    {
        ret_flags &= ~PAGE_PRESENT;
    }

    if (flags & HAL_MEM_WRITE)
    {
        ret_flags |= PAGE_WRITABLE;
    }

    if (flags & HAL_MEM_EXEC)
    {
        ret_flags &= ~PAGE_NO_EXECUTE;
    }

    if (flags & HAL_MEM_USER)
    {
        ret_flags |= PAGE_USER;
    }

    if (flags & HAL_MEM_HUGE)
    {
        ret_flags |= PAGE_HUGE;
    }

    return ret_flags;
}

static Res paging_get_pml_alloc(uintptr_t *pml, size_t index, bool alloc)
{
    if ((pml[index] & PAGE_PRESENT))
    {
        return uok$(hal_mmap_l2h(PAGE_GET_PHYS(pml[index])));
    }
    else if (alloc)
    {
        PhysObj obj = pmm_alloc(1);
        uintptr_t ptr_hddm = hal_mmap_l2h(obj.base);
        memset((void *)ptr_hddm, 0, obj.len);

        pml[index] = obj.base | PAGE_PRESENT | PAGE_WRITABLE | PAGE_USER;

        return uok$(ptr_hddm);
    }

    return err$(RES_NOMEM);
}

static Res kmmap_page(uintptr_t *pml, uint64_t virt, uint64_t phys, int64_t flags)
{
    if (phys % PMM_PAGE_SIZE != 0 || virt % PMM_PAGE_SIZE != 0)
    {
        return err$(RES_BADALIGN);
    }

    size_t pml1_entry = PMLX_GET_INDEX(virt, 0);
    size_t pml2_entry = PMLX_GET_INDEX(virt, 1);
    size_t pml3_entry = PMLX_GET_INDEX(virt, 2);
    size_t pml4_entry = PMLX_GET_INDEX(virt, 3);

    uintptr_t *pml3 = (uintptr_t *)try$(paging_get_pml_alloc(pml, pml4_entry, true));

    if (page_size == gib$(1) && flags & PAGE_HUGE)
    {
        pml3[pml3_entry] = phys | flags;
        return ok$();
    }

    uintptr_t *pml2 = (uintptr_t *)try$(paging_get_pml_alloc(pml3, pml3_entry, true));

    if (flags & PAGE_HUGE)
    {
        pml2[pml2_entry] = phys | flags;
        return ok$();
    }

    uintptr_t *pml1 = (uintptr_t *)try$(paging_get_pml_alloc(pml2, pml2_entry, true));

    pml1[pml1_entry] = phys | flags;
    return ok$();
}

static Res kmmap_section(uintptr_t start, uintptr_t end, uint8_t flags)
{
    KernelMmap kaddr = loader_get_kernel_mmap();
    int64_t flags_arch = transform_flags(flags);
    size_t end_loop = align_up$(end, PMM_PAGE_SIZE);

    for (size_t i = align_down$(start, PMM_PAGE_SIZE); i < end_loop; i += PMM_PAGE_SIZE)
    {
        uintptr_t phys = i - kaddr.virt + kaddr.phys;
        try$(kmmap_page(pml4, i, phys, flags_arch));
    }

    return ok$();
}

Res hal_space_map(HalPage *self, uintptr_t virt, uintptr_t phys, size_t len, uint8_t flags)
{
    if (phys % PMM_PAGE_SIZE != 0 || virt % PMM_PAGE_SIZE != 0 || len % PMM_PAGE_SIZE != 0)
    {
        return err$(RES_BADALIGN);
    }

    int64_t flags_arch = transform_flags(flags);
    const size_t map_psize = flags & HAL_MEM_HUGE ? page_size : PMM_PAGE_SIZE;

    size_t end = align_up$(len, map_psize);
    size_t aligned_virt = align_down$(virt, map_psize);
    size_t aligned_phys = align_down$(phys, map_psize);

    for (size_t i = 0; i < end; i += map_psize)
    {
        try$(kmmap_page((uintptr_t *)self, aligned_virt + i, aligned_phys + i, flags_arch));
    }

    return ok$();
}

Res hal_space_unmap(HalPage *space, uintptr_t virt, size_t len)
{
    if (virt % PMM_PAGE_SIZE != 0 || len % PMM_PAGE_SIZE != 0)
    {
        return err$(RES_BADALIGN);
    }

    // TODO: Huge pages ?

    for (size_t i = virt; i < len; i += PMM_PAGE_SIZE)
    {
        size_t pml1_entry = PMLX_GET_INDEX(i, 0);
        size_t pml2_entry = PMLX_GET_INDEX(i, 1);
        size_t pml3_entry = PMLX_GET_INDEX(i, 2);
        size_t pml4_entry = PMLX_GET_INDEX(i, 3);

        uintptr_t *pml3 = (uintptr_t *)try$(paging_get_pml_alloc((uintptr_t *)space, pml4_entry, false));
        uintptr_t *pml2 = (uintptr_t *)try$(paging_get_pml_alloc(pml3, pml3_entry, false));
        uintptr_t *pml1 = (uintptr_t *)try$(paging_get_pml_alloc(pml2, pml2_entry, false));

        pml1[pml1_entry] = 0;
    }

    return ok$();
}

Res paging_init(void)
{
    PhysObj obj = pmm_alloc(1);
    if (obj.base == 0)
    {
        return err$(RES_NOMEM);
    }

    log$("PML4: %p", obj.base);
    pml4 = (uintptr_t *)hal_mmap_l2h((uintptr_t)obj.base);

    memset((void *)pml4, 0, obj.len);

    if (cpuid_has_1gb_pages())
    {
        log$("1GB pages are supported");
        page_size = gib$(1);
    }
    else
    {
        log$("1GB pages are not supported, defaulting to 2MB pages");
        page_size = mib$(2);
    }

    kmmap_section((uintptr_t)text_start_addr, (uintptr_t)text_end_addr, HAL_MEM_READ | HAL_MEM_EXEC);
    kmmap_section((uintptr_t)rodata_start_addr, (uintptr_t)rodata_end_addr, HAL_MEM_READ);
    kmmap_section((uintptr_t)data_start_addr, (uintptr_t)data_end_addr, HAL_MEM_READ | HAL_MEM_WRITE);

    log$("Kernel sections mapped");

    size_t end = max$(gib$(4), pmm_available_pages() * PMM_PAGE_SIZE);
    uint64_t flags = transform_flags(HAL_MEM_WRITE | HAL_MEM_READ | HAL_MEM_HUGE);

    for (size_t i = page_size; i < end; i += page_size)
    {
        try$(kmmap_page(pml4, hal_mmap_l2h(i), i, flags));
    }

    HandoverPayload *hand = handover();
    HandoverRecord rec;

    handover_foreach_record(hand, rec)
    {
        if (rec.tag != HANDOVER_FB)
        {
            try$(hal_space_map((HalPage *)pml4,
                               align_down$(hal_mmap_l2h(rec.start), PMM_PAGE_SIZE),
                               align_down$(rec.start, PMM_PAGE_SIZE),
                               align_up$(rec.size, PMM_PAGE_SIZE),
                               HAL_MEM_READ | HAL_MEM_WRITE | HAL_MEM_HUGE));
        }
    }

    log$("Memory mapped");
    hal_space_apply((HalPage *)pml4);
    log$("Space applied");

    return ok$();
}

void hal_space_apply(HalPage *space)
{
    asm_write_cr(3, hal_mmap_h2l((uintptr_t)space));
}

Res hal_space_create(HalPage **self)
{

    PhysObj obj = pmm_alloc(1);
    uintptr_t *space = (uintptr_t *)hal_mmap_l2h(obj.base);
    memset((void *)space, 0, obj.len);

    if (obj.base == 0)
    {
        return err$(RES_NOMEM);
    }

    memset((void *)space, 0, PMM_PAGE_SIZE);

    for (size_t i = 255; i < 512; i++)
    {
        space[i] = pml4[i];
    }

    *self = (HalPage *)space;
    return ok$();
}

HalPage *hal_space_kernel(void)
{
    return (HalPage *)pml4;
}

Res hal_virt2phys(HalPage *space, uintptr_t virt)
{
    size_t pml1_entry = PMLX_GET_INDEX(virt, 0);
    size_t pml2_entry = PMLX_GET_INDEX(virt, 1);
    size_t pml3_entry = PMLX_GET_INDEX(virt, 2);
    size_t pml4_entry = PMLX_GET_INDEX(virt, 3);

    uintptr_t *pml3 = (uintptr_t *)try$(paging_get_pml_alloc((uintptr_t *)space, pml4_entry, false));
    uintptr_t *pml2 = (uintptr_t *)try$(paging_get_pml_alloc(pml3, pml3_entry, false));
    uintptr_t *pml1 = (uintptr_t *)try$(paging_get_pml_alloc(pml2, pml2_entry, false));

    if (!(pml1[pml1_entry] & PAGE_PRESENT))
    {
        return err$(RES_NOENT);
    }

    return uok$(PAGE_GET_PHYS(pml1[pml1_entry]));
}

File: src/kernel/archs/x86_64/paging.zig
// Navy - An experimental voyage, one wave at a time.
// Copyright (C) 2025   Keyb <contact@keyb.moe>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

const as = @import("./asm.zig");
const cpuid = @import("./cpuid.zig");

const std = @import("std");
const MapFlag = kutils.MapFlag;
const bootPayload = @import("./boot/root.zig").bootPayload;
const handover = @import("handover");
const kutils = @import("root").kernelUtils;
const logger = @import("logger");
const pmm = @import("root").pmm;
const utils = @import("utils");

var kernelPage: Space = undefined;
var pSize: usize = utils.mem.mib(2);
const log = std.log.scoped(.paging);

const PagingError = error{ AddressNotAligned, LengthNotAligned, PageNotFound, CantReadKernelAddr };

const Sections = struct {
    extern const request_start_addr: u64;
    extern const request_end_addr: u64;

    extern const text_start_addr: u64;
    extern const text_end_addr: u64;

    extern const rodata_start_addr: u64;
    extern const rodata_end_addr: u64;

    extern const data_start_addr: u64;
    extern const data_end_addr: u64;
};

pub const Space = struct {
    root: [*]u64,
    alloc: std.mem.Allocator,

    const PageField = struct {
        const present: u64 = 1 << 0;
        const writable: u64 = 1 << 1;
        const user: u64 = 1 << 2;
        const writeThrough: u64 = 1 << 3;
        const no_cache: u64 = 1 << 4;
        const accessed: u64 = 1 << 5;
        const dirty: u64 = 1 << 6;
        const huge: u64 = 1 << 7;
        const global: u64 = 1 << 8;
        const noExecute: u64 = 1 << 63;
    };

    pub fn blank() !Space {
        var pageAlloc = pmm.PageAllocator.new();
        var allocator = pageAlloc.allocator();
        const root = try allocator.alloc(u8, std.heap.pageSize());

        return .{
            .root = @ptrCast(@alignCast(root.ptr)),
            .alloc = allocator,
        };
    }

    pub fn create() !Space {
        const page = try Space.blank();

        for (255..512) |i| {
            page.root[i] = kernelPage.root[i];
        }

        return page;
    }

    pub fn translateFlags(flags: u8) u64 {
        var f: u64 = Space.PageField.present | Space.PageField.noExecute;

        if (flags & MapFlag.none == MapFlag.none) {
            return 0;
        }

        if (flags & MapFlag.read == MapFlag.read) {}

        if (flags & MapFlag.write == MapFlag.write) {
            f |= Space.PageField.writable;
        }

        if (flags & MapFlag.execute == MapFlag.execute) {
            f &= ~Space.PageField.noExecute;
        }

        if (flags & MapFlag.user == MapFlag.user) {
            f |= Space.PageField.user;
        }

        if (flags & MapFlag.huge == MapFlag.huge) {
            f |= Space.PageField.huge;
        }

        return f;
    }

    fn getEntryIndex(virt: u64, comptime level: u8) u64 {
        const shift: u64 = 12 + level * 9;
        return (virt & (0x1ff << shift)) >> shift;
    }

    fn getEntryAddr(addr: u64) u64 {
        return addr & 0x000ffffffffff000;
    }

    fn getEntry(self: Space, index: usize, alloc: bool) !Space {
        if (self.root[index] & Space.PageField.present == Space.PageField.present) {
            return .{
                .root = @ptrFromInt(pmm.lower2upper(Space.getEntryAddr(self.root[index]))),
                .alloc = self.alloc,
            };
        }

        if (!alloc) {
            return error.PageNotFound;
        }

        const page = try Space.blank();
        self.root[index] = pmm.upper2lower(@intFromPtr(page.root)) | Space.PageField.present | Space.PageField.writable | Space.PageField.user;
        return page;
    }

    pub fn mapPage(self: Space, virt: u64, phys: u64, flags: u64) !void {
        std.debug.assert(virt % std.heap.pageSize() == 0);
        std.debug.assert(phys % std.heap.pageSize() == 0);

        const pml4Index = Space.getEntryIndex(virt, 3);
        const pml3Index = Space.getEntryIndex(virt, 2);
        const pml2Index = Space.getEntryIndex(virt, 1);
        const pml1Index = Space.getEntryIndex(virt, 0);

        var pml3 = try self.getEntry(pml4Index, true);

        if (flags & Space.PageField.huge == Space.PageField.huge and pSize == utils.mem.gib(1)) {
            pml3.root[pml3Index] = phys | flags;
            return;
        }

        var pml2 = try pml3.getEntry(pml3Index, true);
        if (flags & Space.PageField.huge == Space.PageField.huge) {
            pml2.root[pml2Index] = phys | flags;
            return;
        }

        var pml1 = try pml2.getEntry(pml2Index, true);
        pml1.root[pml1Index] = phys | flags;
    }

    fn unmapPage(self: Space, virt: u64) !void {
        const pml4Index = Space.getEntryIndex(virt, 3);
        const pml3Index = Space.getEntryIndex(virt, 2);
        const pml2Index = Space.getEntryIndex(virt, 1);
        const pml1Index = Space.getEntryIndex(virt, 0);

        const pml3 = try self.getEntry(pml4Index, false);
        const pml2 = try pml3.getEntry(pml3Index, false);
        const pml1 = try pml2.getEntry(pml2Index, false);

        const entry = &pml1.root[pml1Index];
        entry.* = 0;
    }

    pub fn virt2phys(self: Space, virt: u64) !u64 {
        const pml4Index = Space.getEntryIndex(virt, 3);
        const pml3Index = Space.getEntryIndex(virt, 2);
        const pml2Index = Space.getEntryIndex(virt, 1);
        const pml1Index = Space.getEntryIndex(virt, 0);

        const pml3 = try self.getEntry(pml4Index, false);
        const pml2 = try pml3.getEntry(pml3Index, false);
        const pml1 = try pml2.getEntry(pml2Index, false);

        const entry = pml1.root[pml1Index];
        return Space.getEntryAddr(entry);
    }

    pub fn map(self: Space, virt: u64, phys: u64, len: u64, flags: u8) !void {
        const _align: usize = if (flags & MapFlag.huge == MapFlag.huge) pSize else std.heap.pageSize();

        const aligned_virt = std.mem.alignBackward(u64, virt, _align);
        const aligned_phys = std.mem.alignBackward(u64, phys, _align);
        const aligned_len = std.mem.alignForward(u64, len, _align);

        const f = Space.translateFlags(flags);

        var i: usize = 0;
        while (i < aligned_len) : (i += _align) {
            try self.mapPage(aligned_virt + i, aligned_phys + i, f);
        }
    }

    pub fn unmap(self: Space, virt: u64, len: usize) !void {
        const aligned_virt = std.mem.alignBackward(u64, virt, std.heap.pageSize());
        const aligned_len = std.mem.alignBackward(u64, len, std.heap.pageSize());

        var i: usize = 0;
        while (i < aligned_len) : (i += std.heap.pageSize()) {
            try self.unmapPage(aligned_virt + i);
        }
    }

    pub fn load(self: Space) void {
        as.cr3.write(@truncate(pmm.upper2lower(@intFromPtr(self.root))));
    }
};

fn mapSection(start: u64, end: u64, flags: u8) !void {
    const payload = bootPayload();
    var kernel: handover.Record = undefined;

    for (payload.records[0..payload.count]) |r| {
        if (r.tag == @intFromEnum(handover.Tags.KERNEL)) {
            kernel = r;
        }
    }

    const aligned_start = std.mem.alignBackward(u64, start, std.heap.pageSize());
    const len = std.mem.alignForward(u64, end - start, std.heap.pageSize());

    try kernelPage.map(
        aligned_start,
        (aligned_start - handover.KERNEL_BASE) + kernel.start,
        len,
        flags,
    );
}

pub fn kernelSpace() Space {
    return kernelPage;
}

pub fn setup() !void {
    kernelPage = try Space.blank();
    log.debug("Kernel page allocated at {x:0>16}", .{@intFromPtr(kernelPage.root)});

    if (try cpuid.has1GBPages()) {
        pSize = utils.mem.gib(1);
        log.debug("1GB pages supported", .{});
    }

    try mapSection(
        @intFromPtr(&Sections.request_start_addr),
        @intFromPtr(&Sections.request_end_addr),
        MapFlag.read,
    );

    try mapSection(
        @intFromPtr(&Sections.text_start_addr),
        @intFromPtr(&Sections.text_end_addr),
        MapFlag.read | MapFlag.execute,
    );

    try mapSection(
        @intFromPtr(&Sections.rodata_start_addr),
        @intFromPtr(&Sections.rodata_end_addr),
        MapFlag.read,
    );

    try mapSection(
        @intFromPtr(&Sections.data_start_addr),
        @intFromPtr(&Sections.data_end_addr),
        MapFlag.read | MapFlag.write,
    );

    log.debug("Kernel sections mapped", .{});

    try kernelPage.map(
        pmm.lower2upper(pSize),
        pSize,
        @max(utils.mem.gib(4), pmm.availableMem()),
        MapFlag.write | MapFlag.read | MapFlag.huge,
    );

    const payload = bootPayload();
    for (payload.records[0..payload.count]) |r| {
        if (kutils.handover.isRecordMappable(r)) {
            try kernelPage.map(
                pmm.lower2upper(r.start),
                r.start,
                r.size,
                MapFlag.read | MapFlag.write | MapFlag.huge,
            );
        } else if (r.tag == @intFromEnum(handover.Tags.STACK)) {
            try kernelPage.map(
                r.start + handover.KERNEL_BASE,
                r.start,
                r.size,
                MapFlag.read | MapFlag.write,
            );
        }
    }

    log.debug("Kernel pages mapped", .{});
    kernelPage.load();
    log.debug("Paging initialized", .{});
}
